### DISCUSSION TITLE
Allow for the generation of SVG rather than HTML

### DISCUSSION THREAD

**Discussion by dysbulic** [vercel/next.js#23907] [Ideas]
### Describe the feature you'd like to request

I would like to be able to use Next.js to generate SVG images. Specifically, I would like to be able to nest images using the `image` tag, so it will not work if the images are wrapped in HTML.

### Describe the solution you'd like

Currently, [NextDocument only generates HTML](//github.com/vercel/next.js/blob/canary/packages/next/pages/_document.tsx).

I would like to see the common generation code remain in NextDocument, and on top if that two parallel classes: Document and SVGDocument. When subclassing SVGDocument the `Main` and `NextScript` elements would remain the same, but `Html` would be replaced with `Svg`, and `Head` simply wouldn't be used b/c SVG `defs` elements are allowed anywhere in the document.

### Describe alternatives you've considered

It is theoretically possible that I could use a `foreignObject` tag to embed HTML pages into my SVG. Doing so, however, creates a bunch of sizing and rendering issues that could be eliminated if SVG were generated directly.

**Comment by Janpot** [comment]
Would an API route satisfy your use case?
```tsx
// pages/api/[var].js
import { renderToStaticMarkup }  from 'react-dom/server'

export default (req, res) => {
  res.setHeader('content-type', 'image/svg+xml')
  res.status(200).send(renderToStaticMarkup(
    <svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 100 100'>
      <rect width={100} height={100} fill='#F00'/>
      <text x={50} y={68} font-size={48} color='#000' text-anchor='middle'>
        {req.query.var}
      </text>
    </svg>
  ))
}
```
This will allow you to generate svg using react, you could use it on your pages as:
```ts
// pages/index.js
<img src='/api/hello' width={200} height={200} />
```

**Reply by dysbulic** [reply]
Recognizing how deeply HTML is baked into the current rendering structure, this seems like it may be the best alterative.

It would be really useful someday to have SVG as a first-class rendering option, but in the interim this is a useful alternative.

Thanks for the example.
